### ISSUE TITLE
How to test if customer app uses pristine Spartacus libraries

### ISSUE THREAD

**Issue by Xymmer** [SAP/spartacus#6228]
Our support team needs this:
How we can tell if Spartacus libraries being used by customers were modified.
i.e. how we can provide a tool or procedure that tests if Spartacus libraries being used by a customer are “pristine” - same as from npmjs.

I assume this can only be done with customer source code and not on published Spartacus-based storefront, or could it be done with published storefront?

This is important as when we get a support request, the first thing we do is make sure the libraries are not modified. We also recommend to support that they update the customer to latest release.

**Comment by LTiger14** [comment]
I would say do the following:

**Comment by LTiger14** [comment]
I would say do the following:

1. Backup your `node_modules` (copy the folder to another location)
2. Delete your `node_modules`
3. `yarn install` (or npm install)
4. See if the issue persists

**Comment by Xymmer** [comment]
yes, but we can save time by running a small app that checks the libs right? like runs a hash or does a compare?

**Comment by LTiger14** [comment]
Yes that would probably be possible.

**Comment by hackergil** [comment]
We can compute the SHA checksum of each one of the libraries. If the libs have been changed the SHA won't match.

**Comment by reyhanehsalkhi** [comment]
I have put the steps to run the jar file in the below page 

https://cxwiki.sap.com/display/~reyhaneh.salkhi@hybris.com/How+to+test+if+customer+app+uses+pristine+Spartacus+libraries

**Comment by hackergil** [comment]
@reyhanehsalkhi where can I review the code for this?

**Comment by reyhanehsalkhi** [comment]
@hackergil thanks for the comments. 
I have put the source code on github.

https://github.com/reyhanehsalkhi/support

**Comment by hackergil** [comment]
@reyhanehsalkhi I have created an internal git repo for this and moved your files there
https://github.tools.sap/cx-commerce/spartacus-tools

Please add the following:
- A build file to build the tool so that we can distribute it as a jar file and run the tests in an easy way (maven or gradle).
- Instructions of how to build, test and run the tool (in the README file)
- Unit tests for the classes that are not tested.

**Comment by reyhanehsalkhi** [comment]
@hackergil  I have  added the requirements and pushed the code to master branch
https://github.tools.sap/cx-commerce/spartacus-tools

thanks

**Comment by hackergil** [comment]
Hi Rey. General feedback about the latest code changes, here are my comments:

General project structure

- From now on, please do a pull request instead of checking in directly to master. This makes easier to review the changes and merge atomic commits into master.
- Please add a .gitignore file to avoid having the following under source control:
  - Hidden mac files (.DS_Store)
  - target folder (generated classes go here)
  - IDE files (eclipse, intellj, vscode, etc)
  - eclipse files (.classpath, .project, .settings)

A couple of questions:

 - What is the file dependency-reduced-pom.xml for?
- I believe our java version for this tool should be 11, not 8. Any reason why we are using an old java version?

**Comment by Xymmer** [comment]
Saw the demo, was great!

Had some thoughts on the UI, here's what I'd like to see:
0. Assumption: Putting ourselves in the boots of the support, he/she likely will tell customer to send their source. So support person has their app which includes the libraries.
1. Support person downloads Rey's validation tool and puts it into customer's node_modules/@spartacus
2. Support runs the tool. 
3. The tool looks for folders named assets, cds, etc. and grabs the versions from package.json.
4. The tool automatically downloads the version of the libraries from npmjs.org and runs the comparisons.
5. A report is displayed:
- Display all known libraries and whether they are pristine or they didn't exist.
- Display a warning if all the libraries aren't the same versions.

The tl;dr is that we can't expect the support people to be on top of all libraries all the time so we should build some intelligence into the tool.

Example report:
----
- assets (1.4.1): No differences
- core (1.4.1): DIFFERENCES FOUND
- cds: Not found in customer source code
- schematics (1.4.0): Same
- ...
- Unrecognized folders found: None

Warning: Not all libraries are using same patch version.
